pseudopotential_calculator/calculations/slab.py

```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
     return np.abs(cast(float, position_0) - cast(float, position_1))
 
 
+def _get_constraints(
+    atoms: Atoms,
+    n_free_layer: int,
+) -> list[FixAtoms | FixedLine]:
+    heights = _get_heights_sorted(atoms)
+    sorted_indices = np.argsort(heights)
+
+    free_layer_indices = [i for i in sorted_indices if i < n_free_layer]
+    fixed_layer_indices = [i for i in sorted_indices if i >= n_free_layer]
+    return [
+        FixAtoms(fixed_layer_indices),
+        FixedLine(free_layer_indices, [0, 0, 1]),  # type: ignore bad lib
+    ]
+
+
 def get_surface(
     atom: Atoms,
     slab_direction: tuple[int, int, int],
     *,
-    n_layer: int,
-    n_vacuum_layer: int,
+    n_fixed_layer: int = 0,
+    n_free_layer: int = 0,
+    n_vacuum_layer: int = 0,
 ) -> Atoms:
-    slab = cast(Atoms, surface(atom, slab_direction, n_layer, 0))
+    n_layer = n_free_layer + n_fixed_layer
+    atoms = cast(Atoms, surface(atom, slab_direction, n_layer, 0))
 
     height_per_layer = _get_height_per_layer(atom, slab_direction)
-    add_vacuum(slab, n_vacuum_layer * height_per_layer)
-    return slab
+    add_vacuum(atoms, n_vacuum_layer * height_per_layer)
+
+    atoms.set_constraint(  # type: ignore bad lib
+        _get_constraints(atoms, n_free_layer),
+    )
+    return atoms
 
 
-def get_n_vacuum_layers_from_out_put(atom: Atoms) -> float:
+def _get_n_vacuum_layers_from_slab(atom: Atoms) -> float:
     cell_height = atom.get_cell().array[-1][-1]
 
-    atom_heights = list[float](atom.positions[:, 2])  # type: ignore unknown
-    slab_height = max(atom_heights) - min(atom_heights)
-    height_per_layer = slab_height / len(atom_heights)
+    slab_height = get_n_th_slab_layer_height(atom, 1)
+    height_per_layer = get_height_per_slab_layer(atom)
 
     vacuum_height = cell_height - slab_height
 
     return vacuum_height / height_per_layer
 
 
+def _get_heights_sorted(atom: Atoms) -> np.ndarray[Any, np.dtype[np.float64]]:
+    heights = cast(np.ndarray[Any, np.dtype[np.float64]], atom.positions[:, 2])  # type: ignore bad lib
+    return np.unique(heights)[::-1]
+
+
+def get_n_th_slab_layer_height(atom: Atoms, n_th_layer: int) -> float:
```

Comment:
get_height_of_layer(atoms: Atoms, layer_idx: int) -> float:

Could you also replace all atom: Atoms with atoms: Atoms ...